lib/src/deploy.rs

```diff
@@ -340,6 +347,64 @@ pub(crate) fn switch_origin_inplace(root: &Dir, imgref: &ImageReference) -> Resu
     Ok(newest_deployment)
 }
 
+pub fn get_kargs(repo: &ostree::Repo, fetched: &ImageState) -> Result<Vec<String>> {
+    let cancellable = gio::Cancellable::NONE;
+
+    // Get the running kernel commandline arguments
+    let kargs = ostree::KernelArgs::new();
+    ostree::KernelArgs::append_proc_cmdline(
+        &kargs,
+        cancellable,
+    )?;
+    let mut kargs: Vec<String> = kargs.to_strv().iter().map(|s| { s.as_str().to_string() }).collect();
+
+    // Get the kargs in kargs.d of the booted system
+    let mut existing_kargs = vec![];
+    let fragments = liboverdrop::scan(&["/usr/lib"], "bootc/kargs.d", &["toml"], true);
+    for (_name, path) in fragments {
+        let buffer = std::fs::read_to_string(&path)?;
+        existing_kargs.push(buffer.trim().to_string());
+    }
+    
+    // Get the kargs in kargs.d of the remote image
+    let mut remote_kargs = vec![];
+    let (fetched_tree, _) = repo.read_commit(fetched.ostree_commit.as_str(), cancellable)?;
+    let fetched_tree = fetched_tree.resolve_relative_path("/usr/lib/bootc/kargs.d");
+    let fetched_tree = fetched_tree.downcast::<ostree::RepoFile>().expect("downcast");
+    match fetched_tree.query_exists(cancellable) {
+        true => {}
+        false => {
+            return Ok(vec![]);
+        }
+    }
+    let queryattrs = "standard::name,standard::type";
+    let queryflags = gio::FileQueryInfoFlags::NOFOLLOW_SYMLINKS;
+    let fetched_iter = fetched_tree.enumerate_children(queryattrs, queryflags, cancellable)?;
+    while let Some(fetched_info) = fetched_iter.next_file(cancellable)? {
+        let fetched_child = fetched_iter.child(&fetched_info);
+        let fetched_child = fetched_child.downcast::<ostree::RepoFile>().expect("downcast");
+        fetched_child.ensure_resolved()?;
+        let fetched_contents_checksum = fetched_child.checksum();
+        let f = ostree::Repo::load_file(repo, fetched_contents_checksum.as_str(), cancellable)?;
+        let file_content = f.0;
+        let mut buffer = vec![];
+        let mut reader = ostree_ext::prelude::InputStreamExtManual::into_read(file_content.unwrap());
+        let _ = std::io::Read::read_to_end(&mut reader, &mut buffer);
+        let s = std::string::String::from_utf8(buffer)?;
+        remote_kargs.push(s.trim().to_string());
+    }
+
+    // get the diff between the existing and remote kargs
+    let mut added_kargs: Vec<String> = remote_kargs.clone().into_iter().filter(|item| !existing_kargs.contains(item)).collect();
```

Comment:
It may make sense to deal with things in terms of https://docs.rs/ostree/latest/ostree/struct.KernelArgs.html which is actually already structured as a combination of "vector+map" (very minor but this filter is an O(N^2) because we're iterating a vector for each vector element).

Reply:
Since the `KernelArgs` struct is basically a hashmap (at least from what I understand from the API), I guess we could possibly look at the keys of one `KernelArgs` struct and check if it exists in the second KernelArgs to get the diff, which should be O(N). However, it seems as though the API doesn't provide an easy way to access the `key-value` pairs, so I'm not sure if using this struct is the best option or if I missed something from the API.